external-crates/move/crates/move-compiler/src/naming/syntax_methods.rs

```diff
@@ -0,0 +1,514 @@
+            context
+                .env
+                .add_diag(diag!(Declarations::InvalidAttribute, (sloc, msg),));
+            return None;
+        } // SyntaxMethodPrekind_::Assign => match mut_opt {
+          //     Some((loc, true)) => SK::Assign,
+          //     _ => {
+          //         let msg = format!(
+          //         "'{}' is only valid if the first parameter's type is a mutable reference as '&mut'",
+          //         SyntaxAttribute::INDEX,
+          //     );
+          //         let ty_msg = "This type is not a reference";
+          //         context.env.add_diag(diag!(
+          //             Declarations::InvalidAttribute,
+          //             (sloc, msg),
+          //             (*ty_loc, msg)
+          //         ));
+          //         return None;
+          //     }
+          // },
+    };
+    Some(sp(sloc, sk))
+}
+
+//-------------------------------------------------------------------------------------------------
+// Types
+//-------------------------------------------------------------------------------------------------
+
+const INVALID_MODULE_MSG: &str = "Invalid 'syntax' definition";
+const INVALID_MODULE_TYPE_MSG: &str = "This type is defined in a different module";
+
+fn determine_subject_type_name(
+    context: &mut Context,
+    cur_module: &ModuleIdent,
+    ann_loc: &Loc,
+    sp!(loc, ty_): &N::Type,
+) -> Option<TypeName> {
+    match ty_ {
+        N::Type_::Apply(_, type_name, _) => {
+            match &type_name.value {
+                N::TypeName_::Multiple(_) => {
+                    let msg = "Invalid type for syntax method definition";
+                    let mut diag = diag!(Declarations::InvalidSyntaxMethod, (*loc, msg));
+                    diag.add_note("Syntax methods may only be defined for single base types");
+                    context.env.add_diag(diag);
+                    None
+                }
+                N::TypeName_::Builtin(builtin_type_name) => {
+                    if let Some(definer) = context.env.primitive_definer(builtin_type_name.value) {
+                        if cur_module == definer {
+                            Some(type_name.clone())
+                        } else {
+                            context.env.add_diag(diag!(
+                                Declarations::InvalidSyntaxMethod,
+                                (*ann_loc, INVALID_MODULE_MSG),
+                                (*loc, INVALID_MODULE_TYPE_MSG)
+                            ));
+                            None
+                        }
+                    } else {
+                        context.env.add_diag(diag!(
+                            Declarations::InvalidSyntaxMethod,
+                            (*ann_loc, "The subject type is a Move builtin type"),
+                        ));
+                        None
+                    }
+                } // something like fake_natives?
```

Comment:
Nit, we can combine these cases a la https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-compiler/src/naming/resolve_use_funs.rs#L128